### PR TITLE
Adresses issue with equality/hashing on module orderings

### LINKED ISSUE #3347: Equality check and hashing for module orderings is broken
I just witnessed this on the latest master (commit a0913ed719ece74a890e19120670cd79dbffc3aa):
```
julia> R, (x,y,z,w) = polynomial_ring(GF(32003), ["x", "y", "z", "w"]);

julia> F = FreeMod(R, 2)
Free module of rank 2 over Multivariate polynomial ring in 4 variables over GF(32003)

julia> o = lex(gens(F))*lex(gens(R))
lex([gen(1), gen(2)])*lex([x, y, z, w])

julia> oo = lex(gens(F))*lex(gens(R))
lex([gen(1), gen(2)])*lex([x, y, z, w])

julia> o == oo
false

julia> hash(o) == hash(oo)
false
```
It is not clear to me whether this is a recent regression or whether this has always been an issue. In any case, this is a serious problem, because caching of Groebner bases does not work. We should probably fix this soon.

CC: @jankoboehm , @wdecker

### PR DESCRIPTION
**Pull request by RafaelDavidMohr**
Adds `==` and hashing for `ModuleOrdering`s.

I don't know if this is the best solution, I essentially just circumvent what julia does by default when calling `==` on structs (which is to call `===` on every field) by instead calling `==` recursively on every field.

If this is potentially to dangerous, let me know. It does fix https://github.com/oscar-system/Oscar.jl/issues/3347 on my machine.

Tag: @ederc

### CODE REVIEW

**Comment by fieker** [comment]
Can't this be handled by matrices or similar like in the monomial Case?
Is there a canonical normal form for those orderings?

**Comment by thofma** [comment]
Hm, yes, I wonder whether this is correct:
```
mutable struct ModOrdering{T} <: AbsModOrdering
   gens::T
   ord::Symbol
```
If one only checks if `gens` and `ord` are equal, then this does not check equality of orderings.

**Comment by RafaelDavidMohr** [comment]
> Can't this be handled by matrices or similar like in the monomial Case? Is there a canonical normal form for those orderings?

It can but it is more involved, see theorem 2 [here](https://www.sciencedirect.com/science/article/pii/S0747717104000999). I decided against this because it would go substantially beyond the addressed issue to implement this and because, as far as I can tell, we don't support any functionality for such general matrix orderings.

If required, I am however happy to take a look at a canonical matrix form for the orderings we do support, maybe it's easier in this case.

**Comment by RafaelDavidMohr** [comment]
> Hm, yes, I wonder whether this is correct:
> 
> ```
> mutable struct ModOrdering{T} <: AbsModOrdering
>    gens::T
>    ord::Symbol
> ```
> 
> If one only checks if `gens` and `ord` are equal, then this does not check equality of orderings.

I apologize I don't quite understand what you mean. Do you have an example?

**Comment by thofma** [comment]
I was thinking about the following situation: The polynomial ring has one variable and the module is free of rank one. Then `:lex` and `:degrevlex` and whatnot define all the same orderings, but the equality check proposed here should yield `false`?

**Comment by RafaelDavidMohr** [comment]
> I was thinking about the following situation: The polynomial ring has one variable and the module is free of rank one. Then `:lex` and `:degrevlex` and whatnot define all the same orderings, but the equality check proposed here should yield `false`?

Yes, good point. Ok, I will think of something else.

**Comment by RafaelDavidMohr** [comment]
Brief update: I am working on this but this will take a while. We will need canonical representations of module orderings and they will take a while to implement.

**Comment by RafaelDavidMohr** [comment]
I've added equality and hashing by embedding the underlying module of the ordering in question into a new polynomial ring and computing the canonical matrix there, which is then used for equality checking and hashing.

While this is not a unique representation of the ordering, it catches at least all reasonable cases that I could come up with.

File: src/Rings/orderings.jl [hunk 1]
```diff
@@ -1717,33 +1717,89 @@ end
...
+      o.canonical_matrix = canonical_matrix(induced_ring_ordering(o))
+    else
+      o.canonical_matrix = _canonical_matrix_intern(o)
+    end
+  end
+  return o.canonical_matrix
+end
+
+function Base.:(==)(o1::ModuleOrdering, o2::ModuleOrdering)
+  return canonical_matrix(o1) == canonical_matrix(o2)
```
Comment (lgoettgens) [inline src/Rings/orderings.jl:1796]: I think that == should compare the underlying modules of the orderings as well. And then same for the hash: the field M should be used there as well

File: src/Rings/orderings.jl [hunk 2]
```diff
@@ -1717,33 +1717,91 @@ end
...
+end
+
+function Base.:(==)(o1::ModuleOrdering, o2::ModuleOrdering)
+  return o1.M == o2.M && canonical_matrix(o1) == canonical_matrix(o2)
+end
+
+function Base.hash(o::ModuleOrdering, h::UInt)
+  h1 = hash(o.M, h)
+  h2 = hash(canonical_matrix(o), h)
+  return hash(h2, h)
```
Comment (lgoettgens) [inline src/Rings/orderings.jl]:
```suggestion
  h = hash(o.M, h)
  h = hash(canonical_matrix(o), h)
  return h
```
`h1` is not used anywhere. I think you wanted to do something like this

Comment (RafaelDavidMohr) [reply src/Rings/orderings.jl]: that makes more sense, yes. Fixed.

File: src/Rings/orderings.jl (hunk 1 above)
Comment (ederc) [reply src/Rings/orderings.jl:1796]: I'm not sure I completely understand why. If I have two different modules and I want to find out if both share the same ordering, how shall I do that? Note that also for `MonomialOrdering` we do not compare the corresponding rings.

Comment (lgoettgens) [reply src/Rings/orderings.jl:1796]: In that case, we should definitely keep it consistent and revert the change here. Sorry for the double work @RafaelDavidMohr

Comment (RafaelDavidMohr) [reply src/Rings/orderings.jl:1796]: No worries, it's reverted.

**Review by ederc — APPROVED** [review]

**Review by fingolfin — APPROVED** [review]

**Pull request merged**